server/src/main/java/io/unitycatalog/server/persist/utils/UriUtils.java

```diff
@@ -0,0 +1,245 @@
+      if (Files.exists(dirPath)) {
+        try (Stream<Path> walk = Files.walk(dirPath, FileVisitOption.FOLLOW_LINKS)) {
+          walk.sorted(Comparator.reverseOrder())
+              .forEach(
+                  path -> {
+                    try {
+                      Files.delete(path);
+                    } catch (IOException e) {
+                      throw new RuntimeException("Failed to delete " + path, e);
+                    }
+                  });
+        } catch (IOException e) {
+          throw new RuntimeException("Failed to delete " + dirPath, e);
+        }
+      } else {
+        throw new IOException("Directory does not exist: " + dirPath);
+      }
+    }
+    return parsedUri;
+  }
+
+  private static URI updateS3Directory(URI parsedUri, Operation op, AwsCredentials awsCredentials) {
+    String bucketName = parsedUri.getHost();
+    String path = parsedUri.getPath().substring(1); // Remove leading '/'
+    String accessKey = awsCredentials.getAccessKeyId();
+    String secretKey = awsCredentials.getSecretAccessKey();
+    String sessionToken = awsCredentials.getSessionToken();
+    String region = ServerPropertiesUtils.getInstance().getProperty("aws.region");
+
+    BasicSessionCredentials sessionCredentials =
+        new BasicSessionCredentials(accessKey, secretKey, sessionToken);
+    AmazonS3 s3Client =
+        AmazonS3ClientBuilder.standard()
+            .withCredentials(new AWSStaticCredentialsProvider(sessionCredentials))
+            .withRegion(region)
+            .build();
+
+    if (op == Operation.CREATE) {
+      if (!path.endsWith("/")) {
+        path += "/";
+      }
+      if (s3Client.doesObjectExist(bucketName, path)) {
+        throw new BaseException(ErrorCode.ALREADY_EXISTS, "Directory already exists: " + path);
+      }
+      try {
+        // Create empty content
+        byte[] emptyContent = new byte[0];
+        ByteArrayInputStream emptyContentStream = new ByteArrayInputStream(emptyContent);
+
+        // Set metadata for the empty content
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentLength(0);
+        s3Client.putObject(new PutObjectRequest(bucketName, path, emptyContentStream, metadata));
+        LOGGER.debug("Directory created successfully: " + path);
+      } catch (Exception e) {
+        throw new BaseException(ErrorCode.INTERNAL, "Failed to create directory: " + path, e);
+      }
+    }
+
+    if (op == Operation.DELETE) {
+      ObjectListing listing;
+      ListObjectsRequest req = new ListObjectsRequest().withBucketName(bucketName).withPrefix(path);
+      do {
+        listing = s3Client.listObjects(req);
+        listing
+            .getObjectSummaries()
+            .forEach(
+                object -> {
+                  s3Client.deleteObject(bucketName, object.getKey());
+                });
+        req.setMarker(listing.getNextMarker());
+      } while (listing.isTruncated());
+    }
+    return parsedUri;
```

Comment:
Note to self: Difference from File Utils: Returning `parsedUri` instead of constructing it again for both operations.